Plutarch/Internal/Lift.hs

```diff
@@ -214,51 +285,83 @@ instance DerivePlutusType (DeriveDataPLiftable a h) where
+{- | @via@-deriving helper, indicating that @wrapper@ has a Haskell-level equivalent
+@h@ by way of encoding of @inner@. It requires that @AsHaskell inner@ has the same
+Haskell representation as @h@
 
 @since WIP
 -}
-unsafeFromUni ::
-  forall (a :: S -> Type) (h :: Type) (s :: S).
-  PLC.DefaultUni `Includes` h =>
-  h ->
-  PLifted a s
-unsafeFromUni =
-  PLifted . punsafeConstantInternal . PLC.someValue @h @PLC.DefaultUni
+newtype DeriveNewtypePLiftable (wrapper :: S -> Type) (inner :: S -> Type) (h :: Type) (s :: S)
+  = DeriveNewtypePLiftable (wrapper s)
+  deriving stock (Generic)
+  deriving anyclass (PlutusType)
+
+-- | @since WIP
+instance DerivePlutusType (DeriveNewtypePLiftable w i h) where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since WIP
+instance (PLiftable inner, Coercible (AsHaskell inner) h) => PLiftable (DeriveNewtypePLiftable wrapper inner h) where
+  type AsHaskell (DeriveNewtypePLiftable wrapper inner h) = h
+  type PlutusRepr (DeriveNewtypePLiftable wrapper inner h) = PlutusRepr inner
+
+  {-# INLINEABLE toPlutarchRepr #-}
+  toPlutarchRepr = toPlutarchRepr @inner . coerce @h @(AsHaskell inner)
+
+  {-# INLINEABLE toPlutarch #-}
+  toPlutarch = punsafeCoercePLifted . toPlutarch @inner . coerce @h @(AsHaskell inner)
 
-{- | Helper for writing 'PLiftable' instances. This is /highly/ unsafe, and only
-suitable for internal use!
+  {-# INLINEABLE fromPlutarchRepr #-}
+  fromPlutarchRepr = coerce . fromPlutarchRepr @inner
+
+  {-# INLINEABLE fromPlutarch #-}
+  fromPlutarch p = coerce . fromPlutarch @inner $ punsafeCoercePLifted p
+
+{- | Similar to 'Identity', but at the level of Plutarch. Only needed when
+writing manual instances of 'PLiftable', or if you want to use 'toPlutarch'
+and 'fromPlutarch' directly.
 
 @since WIP
 -}
-unsafeToUni ::
-  forall (a :: S -> Type) (h :: Type).
-  PLC.DefaultUni `Includes` h =>
-  (forall (s :: S). PLifted a s) ->
-  Either LiftError h
-unsafeToUni t =
-  case compile (Tracing LogInfo DoTracing) (unPLifted t) of
-    Left err -> Left . CouldNotCompile $ err
-    Right compiled -> case evalScriptHuge compiled of
-      (evaluated, _, _) -> case evaluated of
-        Left err -> Left . CouldNotEvaluate $ err
-        Right (Script (UPLC.Program _ _ term)) -> case readKnownConstant term of
-          Left err -> Left . TypeError $ err
-          Right res -> pure res
+type role PLifted nominal nominal
+
+newtype PLifted s a = PLifted {unPLifted :: Term s POpaque}
+
+-- | @since WIP
+punsafeCoercePLifted :: PLifted s a -> PLifted s b
+punsafeCoercePLifted (PLifted t) = PLifted t
+
+-- | @since WIP
+getPLifted :: PLifted s a -> Term s a
+getPLifted (PLifted t) = punsafeCoerce t
+
+-- | @since WIP
+mkPLifted :: Term s a -> PLifted s a
+mkPLifted t = PLifted (popaque t)
+
+{- |  Use this as 'PlutusRepr' when defining 'PLiftable' instance for Scott encoded type
+
+@since WIP
+-}
+newtype PLiftedScott (a :: S -> Type) = PLiftedScott {unPLiftedClosed :: forall (s :: S). PLifted s a}
```

Comment:
This should be just `PClosedLifted` It's not only for scott. We'll probably have SOP support this

Reply:
Yeah, SOP's implementation should be very similar to scott one so this function should be named more generally